Major overhaul of JSON model importer for stability, texture resolution, and tweaks to UV

BolWaffy opened the pull request:
Significantly overhauled how `mcmodel.py` handles textures, materials, and UVs when importing Minecraft JSON models. It fixes several issues with texture lookup, `#` reference resolution, different resource packs, individual JSON model imports, incorrect material indices, etc.

Improves importer reliability across various resource packs.

### **Key Improvements**

* **Reliable texture reference resolution**
  Introduces `get_final_texture_key` to correctly follow chained `#` texture references and avoid circular recursion.

* **Improved image lookup**
  `locate_image` now normalizes paths and searches in this order:

  1. The model’s own resource pack
  2. The active resource pack
  3. The addon’s default fallback

  This fixes many “missing textures” produced by the old version.

* **Correct material creation & indexing**
  Materials are only created for valid, resolved textures.
  Faces now receive consistent `material_index` values, fixing "index out of bounds" errors and other material assignment bugs.

* **Improved UV scaling for non-square textures**
  UV adjustments/scaling only occur when the texture’s **height < width**, preventing distortions on models using 64×32, 128×64, or other non-1:1 textures. UV scaling is not needed when **height > width**.

* **Better error handling & logging**
    * Added better error handling using `try/except` inside the element and face creation loops of `add_model`. This allows the script to **skip problematic model faces or elements** and continue importing the rest of the geometry, preventing the object from completely not importing.

---

I tested on 12 different resource packs and custom mod assets.

If you want to quickly test the new importer. Here is a script that automatically imports and validates each JSON object in a directory.
[Auto Test Minecraft Model Importer.py](https://github.com/user-attachments/files/23691454/Better.Auto.Test.Minecraft.Model.Importer.py)

BolWaffy committed:
Major Overhaul of Model Importer for Stability, Texture Resolution, and tweaks to UV

Significantly overhauled how the JSON importer handles texture references and path resolution, and added stability improvements and new feature support. Also fixed bug for textures with non-square (1:1) dimensions

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -179,33 +175,142 @@ def add_get_material(
-		return os.path.realpath(os.path.join(directory, local_path) + ".png")
+	# If the key is not in textures, return the key itself as the final key.
+	# This handles cases where a face references a key not explicitly defined.
+	return key
```
If `key` is the alternative return value regardless, then the `else` clause above could be removed.

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -179,33 +175,142 @@ def add_get_material(
+	if key in visited:
+		# Detected circular dependency and break recursion (e.g., '#a' -> '#b' -> '#a')
+		env.log(f"Circular texture reference detected for key: {key}")
+		return None
```
Rather than return `None`, this function should return an `MCprepError` object (defined in `conf.py`), since (presumably) circular dependencies are an error with the model.

This would also mean changing the type signature of the function from `Optional[str]` to `Union[str, MCprepError]`.

Here's some reference on how to use `MCprepError` (note that the first parameter should be whatever exception class would be closest to describing the error):
https://github.com/Moo-Ack-Productions/MCprep/blob/4938a3036018feeacf2c9078f2b02784c99f2ffc/MCprep_addon/world_tools.py#L714-L715

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -179,33 +175,142 @@ def add_get_material(
+	local_path_clean = local_path_clean.lstrip(os.sep) # No leading separator
+	if local_path_clean.lower().endswith(".png"):
+		local_path_clean = local_path_clean[:-4]
+	texture_path_fragment = local_path_clean + ".png"
```
I wouldn't use functions from `os` for this. We encourage the use of `pathlib` in new code for working with paths, since it handles a lot of this anyway.

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -179,33 +175,142 @@ def add_get_material(
+				break
+
+		if pack_root:
+			path_model = pack_root / "assets" / namespace / "textures" / texture_path_fragment
```
For this, [`joinpath`](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.joinpath) would be preferable since it's more clear what's going on.

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -179,33 +175,142 @@ def add_get_material(
+		path_active = os.path.normpath(path_active)
+
+		if os.path.isfile(path_active):
+			return path_active
```
As noted earlier, we encourage the use of `pathlib` in new code, since it tends to be a lot cleaner than `os.path`.

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -179,33 +175,142 @@ def add_get_material(
+	except Exception as e:
+		env.log(f"Error checking default minecraft Resource Pack: {e}", vv_only=True)
+	# If neither found, return the path that would be expected in the Active Resource Pack (B) for error messages/debugging.
+	return path_active
```
Since this is supposed to be an error case, `MCprepError` should be used here, to make it clear to any part of the code calling this function what is and isn't an error.

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -179,33 +175,142 @@ def add_get_material(
+		path_active = os.path.normpath(path_active)
+
+		if os.path.isfile(path_active):
+			return path_active
```
Being new code, this should use `pathlib`.

StandingPadAnimations reviewed (commented):
Did a basic skim of the code since this is still a draft. Will probably do something more in-depth this week now that university is out.

As an aside, it would be preferable if you had smaller, more focused commits as opposed to one large commit. It's easier to review stuff when they're smaller commits, since I (and anyone else reviewing this) can more easily review individual changes in isolation; a large commit with multiple changes is harder to review.

BolWaffy committed:
Fix JSON encoding handling when reading model JSONs

Prevents `JSONDecodeErrors` when reading files saved with a BOM

BolWaffy committed:
Add helper functions for path normalization and pack root discovery

- Added `find_all_pack_roots` to handle normal packs, loose/unpacked mod folder/resource packs, and nested packs
- Added `normalize_texture_path` to normalize texture paths, to handle ANY weird, but acceptable, texture paths

BolWaffy committed:
Refactor get_final_texture_key for improved readability

Refactor get_final_texture_key function for clarity and error handling.

BolWaffy committed:
Refactor locate_image function, with new search order

- Added search order: Pack Roots -> Active Pack -> Default Pack
- Added support for relative texture paths starting with `.`
- Fixed handling of namespaces

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -323,7 +323,7 @@ def read_model(
 	try:
 		with open(model_filepath, 'r') as f:
-			obj_data = json.load(f)
+			obj_data = json.load(open(model_filepath, encoding="utf-8-sig"))
```
This additional `open` call is unneeded (and could cause issues with resource cleanup), the encoding can be passed to the `open` call in the `with` statement.

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -175,6 +175,102 @@ def add_get_material(
 	return mat
 
+def find_all_pack_roots(
+	start_path):
```
This function needs a type signature to make it clear what it takes and what it returns

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -175,6 +175,102 @@ def add_get_material(
+		• assets/<namespace>/...
+		OR
+		• textures/	  (for packs missing the assets folder)
+		• models/	  (same case)
```
Bullet points in code files should use `-` and not `•`. `-` is guaranteed to render in any text editor for the most part, while `•` isn't, and could appear as an empty box for devs who use terminal-based text editors (like myself)

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -175,6 +175,102 @@ def add_get_material(
+
+		# Case C: even looser: any folder with a textures/ directory
+		if (parent / "textures").is_dir():
+			roots.append(parent)
```
Case B and C should use `elif`, so that if an earlier case succeeds, the runtime can just skip everything else

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -175,6 +175,102 @@ def add_get_material(
+
+	# 2. Remove redundant "./"
+	while "/./" in path:
+		path = path.replace("/./", "/")
```
[`str.replace`](https://docs.python.org/3/library/stdtypes.html#str.replace) will change all occurrences of the substring, so the `while` loop is unneeded

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -175,6 +175,102 @@ def add_get_material(
+	"""
+
+	if not isinstance(path, str):
+		return path
```
Since the function has been annotated with `path` being a string, this check is unneeded. If someone calls the function with `path` not as a string, pretty much every LSP or static analyzer will catch it anyway

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -175,6 +175,102 @@ def add_get_material(
+	# Make unique and preserve order
+	seen = set()
+	final = []
+	for r in roots:
+		if r not in seen:
+			seen.add(r)
+			final.append(r)
```
More of a style comment, but the condition should be reversed to avoid excessive indentation.

StandingPadAnimations replied:
One thing I have realized here is if `roots` is empty, then `final` will also be empty. That case should be handled as well

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -175,6 +175,102 @@ def add_get_material(
+	roots = []
+	start = Path(start_path).resolve()
+
+	for parent in [start] + list(start.parents):
```
This seems a bit ugly, and I don't understand why both `start` and `start.parents` are both being iterated over?

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -271,41 +271,39 @@ def normalize_texture_path(
-		env.log(f"Circular texture reference detected for key: {key}")
-		return None
-
+		# env.log(f"Circular texture reference detected for key: {key}")
```
This should be removed if it's ultimately going to stay commented

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -271,41 +271,39 @@ def normalize_texture_path(
 
-def get_final_texture_key(
-	texture_ref: str, textures: Dict[str, str], visited: Optional[set] = None) -> Optional[str]:
+def get_final_texture_key(texture_ref, textures, visited=None):
```
Why was the type signature removed? All that needed to change was the return signature to `Union[str, MCprepError]`. Without it, it's ambiguous what type of arguments this function expects, and what values it'll return.

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -304,107 +304,115 @@ def get_final_texture_key(texture_ref, textures, visited=None):
 
-def locate_image(
-	context: Context, textures: Dict[str, str], img: str, model_filepath: Path) -> str:
+def locate_image(context, textures, img, model_filepath):
```
Again, why was the type signature removed here, when only the return signature needed to change?

TheDuckCow commented:
Just chiming in to say thanks @BolWaffy for the continued contributions, and appreciate you Standingpad for the review upkeep

BolWaffy commented:
To answer your question directly: Yes, I do use AI tools, but I keep usage within the guidelines.
I use them primarily as a productivity aid for polishing commit messages, rewording docstrings/comments for clarity, and handling boilerplate or verifying API references (along with me checking the actual documentation). All code logic, decisions, and solutions are my own, and I verify everything I commit.
Regarding the "weird things" you noticed in the subsequent changes: I realize now that I opened this Draft PR earlier than I should have. You likely saw some "scratchpad" commits where I was still working on the implementation and cleanup, rather than polished final code.
As this is my first contribution to a shared project of this scale, I am still tuning my workflow regarding commits and PR timing. Going forward, I will make sure the final code is clean before marking it ready for review.
Thanks for the heads-up!

BolWaffy committed:
Clean up

Cleaned up comments, fixed and formatted Doc type signatures, removed unused imports (traceback) and commented out code

BolWaffy committed:
Code cleanup and fix

Fixed and cleaned up code bits

BolWaffy committed:
MCprepError refactor 

A similar approach to MCprepError for handling errors/warnings. If an error is encountered (e.g. "Texture Recursion"), gracefully handle it and continue processing the next face/element. This prevents the BMesh from becoming corrupt or incomplete, while still providing the user with the error message.

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -277,12 +277,10 @@ def get_final_texture_key(texture_ref: str, textures: Dict[str, str], visited: O
-			f"Circular texture reference detected for key: {key}")
+		(line, file) = env.current_line_and_file()
+		error_collector.collect(MCprepError(Exception("Texture recursion"), line, file, f"Texture recursion on key '{key}' | Using '{key}'"))
+		return key
```
I'm not a fan of this, cause the caller of the function is not aware that this technically failed. Functions should return errors if something fails, and leave it up to the caller of the function to handle however they wish.

Regarding bmesh corruption: if a model has an error in it, in my view, it would be better to abort creating the model entirely, and let the user know, than to silently error in the end.

StandingPadAnimations replied:
Also as an aside, the parenthesis around `line` and `file` is entirely unnecessary. `env.current_line_and_file` returns a tuple, and Python will already interpret `line, file = env.current_line_and_file()` as a tuple unpacking operation.

BolWaffy replied:
I placed parentheses around `line` and `file` since I saw them in the other files, which made it a tad easier on my eyes.
I’ll remove them before taking this PR out of draft.

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -1185,6 +1177,43 @@ def draw_callback(self, context) -> None:
+				self.report({'ERROR'}, msg)
+				
+		self._model_errors = [] # Clear errors after reporting
+		self._unhandled_errors = []
```
As noted earlier, this isn't really a good way to gracefully handle errors. Low level functions should return error objects, and graceful handling should be done in the higher level functions that call them.

Perhaps if you *really* wanted to, you could make low level functions have the return signature of `Union[type_of_return, Tuple[type_of_return, MCprepError], MCprepError]`, where:

- `type_of_return` is the type of whatever object is returned when the function is successful
- `Tuple[type_of_return, MCprepError]` is returned if an issue occurs, but it's really a warning
- `MCprepError` for errors where the spawner needs to stop immediately

Then the caller could check what type was returned using `isinstance` calls and, if an warning or error is returned, handle accordingly. 

Again, IMO any error in the model should abort the model creation process entirely, but that's an option if complete abortion isn't desired

BolWaffy replied:
Ohh, gotcha. That makes more sense now. Thanks for the explanation. I'll rework that.

BolWaffy committed:
Refactor error handling in texture functions

BolWaffy committed:
Removed WIP comments

Removed commented 'Pass Up ERROR' lines

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -289,23 +290,23 @@ def get_final_texture_key(texture_ref: str, textures: Dict[str, str], error_coll
 		# Is another reference
 		if isinstance(value, str) and value.startswith("#"):
-			return get_final_texture_key(value, textures, error_collector, visited)
+			return get_final_texture_key(value, textures, visited)
```
While it's great that the function now returns an error, the return signature doesn't match the actual return type. The return signature is `Tuple[str, Optional[MCprepError]]`, which indicates that all return values will be a tuple with a string, and either an `MCprepError` object, or `None`. However, the return paths here don't match this (the error path returns just an `MCprepError` object, and the main path returns just a string).

I know I mentioned `Tuple[str, Optional[MCprepError]]` as an option for warnings as opposed to complete errors (if that was desired). What that would mean is, in the case of a successful run, the function would return `("some key", None)`, and in the case of a warning, `("some key", MCprepError(...))`. In both cases, the key would be returned, and a warning could also be propagated (if the desired behavior was for the spawner to continue on, and not abort completely).

Personally, I do like the return paths here being "absolute" (i.e. string or error, not both), so my suggestion would be to change the return signature to `Union[str, MCprepError]`, and keep the existing return paths as is. If, however, the desired behavior is to treat error objects as warnings, refer to the above.

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -289,23 +290,23 @@ def get_final_texture_key(texture_ref: str, textures: Dict[str, str], error_coll
-
-def locate_image(context: Context, textures: Dict[str, str], img: str, model_filepath: str, error_collector: "ModelErrorCollector") -> str:
+def locate_image(
+	context: Context, textures: Dict[str, str], img: str, model_filepath: str) -> Union[str, Optional[MCprepError]]:
```
`Optional[MCprepError]` is equivalent to `Union[MCprepError, None]`. Based on the code paths, the `Optional` is unnecessary, and the return signature can be simplified to `Union[str, MCprepError]`

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -481,14 +485,15 @@ def read_model(
 
 def add_model(
-	model_filepath: Path, obj_name: str = "MinecraftModel", error_collector: "ModelErrorCollector" = None) -> Tuple[int, bpy.types.Object]:
+	model_filepath: Path, obj_name: str = "MinecraftModel") -> Union[tuple[int, bpy.types.Object], Optional[MCprepError]]:
```
As note earlier, `Optional[MCprepError]` is a shorthand for `Union[MCprepError, None]`. Since this function only returns an object of either `Tuple[int, bpy.types.Object]` or `MCprepError` the `Optional` can be removed, and `MCprepError` can be kept.

Also, since MCprep supports Blender 2.8 (which uses Python 3.7), the built-in `tuple` type annotation is unfortunately unavailable (that was introduced in Python 3.9). Instead, the `Tuple` annotation from `typing` has to be used (just a simple case of swapping `t` for `T`).

StandingPadAnimations reviewed (commented):
I like the move towards absolute errors, and returning `MCprepError` objects in lower level functions. Just some comments about type annotations that either don't match the actual return values of the function, or could be simplified.

BolWaffy committed:
Update return types in mcmodel.py functions

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -96,14 +94,12 @@ def rotate_around(
 	rot_axis: str = 'y',
-	rot_angle: float = 0
-) -> list:
+	rot_angle: float = 0) -> list:
```
Similar to `tuple`, `list` is a Python 3.9 type annotation, so it has to be replaced with `List` for Python 3.7 support. 

Also, the type of the elements in the list needs to be added, so the return signature would be something like `List[type_of_elements]`

StandingPadAnimations replied:
Ok so I need some clarification, what are the intended types for `faces` and `edges` in this function? I know `verts` is supposed to be a `List[VectorType]`, but I can't seem to determine if `edges` is supposed to be that as well (and `faces` is a list of lists)? Also the return values don't seem to match the signature (at least according to my editor's LSP)

For now, I'll assume the return value is actually supposed to be `Tuple[List[VectorType], List[VectorType], List[Tuple[int, int, int, int]]]`, but correct me if that assumption is wrong

StandingPadAnimations replied:
As an aside, if you haven't already, I would recommend making sure your Python LSP is configured to analyze type annotations. That way, at least your editor will warn you if the types of stuff don't match

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -179,33 +173,228 @@ def add_get_material(
+	roots = []
+	start = Path(start_path).resolve()
+
+	for parent in list(start.parents):
```
`PurePath.parents` is a sequence, so list conversion isn't needed for iteration.

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -179,33 +173,228 @@ def add_get_material(
+	return path
+
+def get_final_texture_key(
+	texture_ref: str, textures: Dict[str, str], visited: Optional[set] = None) -> Union[str, MCprepError]:
```
I'll have to double-check if `set` is also a Python 3.9 annotation, but iirc, this has to be swapped for `Set` (from the `typing` module)

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -813,16 +1094,17 @@ class MCPREP_OT_import_minecraft_model_file(
-		if res:
-			self.report({'ERROR'}, res.msg)
+		result = self.create_and_place_json_model(context, Path(self.filepath))
+		print(result)
```
I'm assuming this is a temporary debug print

StandingPadAnimations reviewed (commented):
Did a final skim over the pull request, just a few small things to change (and I need to test this locally) before we merge this into `dev`.

Since the requested changes are minimal, I'll just push the changes myself (probably some time this later this week).

StandingPadAnimations committed:
fix: Fix type annotations

Signed-off-by: Maryam Sheikh <mahid@standingpad.org>

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -179,33 +173,228 @@ def add_get_material(
+			seen.add(r)
+			final.append(r)
+
+	return final
```
My editor's LSP ([`based-pyright`](https://github.com/DetachHead/basedpyright)) is warning me that `final` is defined as `List[Path]`, while the `find_all_pack_roots` has the return signature of `str`. I'll make the assumption that `find_all_pack_roots` is supposed to return `List[Path]`, and see where that goes.

StandingPadAnimations committed:
Fix: remove unnecessary list conversion

Signed-off-by: Maryam Sheikh <mahid@standingpad.org>

StandingPadAnimations committed:
fix: remove unnecessary print statement

Signed-off-by: Maryam Sheikh <mahid@standingpad.org>

zNightlord commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -124,21 +120,19 @@ def add_element(
-		[7, 6, 2, 3],  # down
-		[4, 0, 3, 7],  # west
-		[1, 5, 6, 2]]  # east
+	edges: List[VectorType] = []
```
This is a list of two indices so `List[Tuple[int, int]]`

StandingPadAnimations replied:
Ah that makes sense, I'll fix that real quick

zNightlord commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -96,14 +94,12 @@ def rotate_around(
 	rot_axis: str = 'y',
-	rot_angle: float = 0
-) -> list:
+	rot_angle: float = 0) -> Tuple[List[VectorType], List[VectorType], List[Tuple[int, int, int, int]]]:
```
Edge is `List[Tuple[int, int]]`

zNightlord commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -323,292 +509,381 @@ def add_model(
+
+					# uv can be rotated 0, 90, 180, or 270 degrees
+					uv_rot = d_face.get("rotation")
+					if uv_rot is None:
```
You can do `d_face.get("rotation", 0)` for default fallback here

StandingPadAnimations replied:
Oddly, my LSP is stating `d_face` is a string, so I'll have to investigate that

zNightlord reviewed (commented):
Looking around it is fine in general, there are weird parts as mentioned probably how AI picture it and goes safe, again it is fine.

If something you check, not exists and purposely created it then it just makes no sense to check it again honestly.
https://github.com/Moo-Ack-Productions/MCprep/blob/ca3594b38905ca37ceb2c50f0ed50ecd9688515c/MCprep_addon/spawner/mcmodel.py#L635-L641

StandingPadAnimations committed:
fix: Fix type annotations for edges in JSON spawner

Signed-off-by: Maryam Sheikh <mahid@standingpad.org>

BolWaffy committed:
Remove unused materials_remap dictionary

Removed unused materials_remap dictionary because of the new get_final_texture_key

StandingPadAnimations commented:
Could you clarify what the types of `faces` and `d_face` are supposed to be in this section? My LSP is stating that `faces` is either a `VectorType`, `Dict[str, str]`, or `None`. I've added a check for `Dict` in the same check for `None`, but this leaves `d_face` as a `str` (thus making `d_face.get(...)` be invalid, since `str` does not have the `get` method), since `faces` would be narrowed down to a `Dict[str, str]`, and `faces.get(...)` would, as such, return a `str`. 

```py
			faces = e.get("faces")
			for i in range(len(element[2])):
				try:
					f = element[2][i]


					if not faces:
						continue


					face_name = FACE_DIRECTIONS[i]
					d_face = faces.get(face_name)
					if not d_face:
						continue


					face_mat = d_face.get("texture")


					if not face_mat:
						continue
```

StandingPadAnimations commented:
It looks like the `e` that's being iterated in the for loop is assumed to be a `Dict[str, Dict[str, str]]` (aliased as `TexFace`). I'll try to refactor this so static analyzers and LSPs are aware of this assumption. @BolWaffy if you haven't already, I highly encourage using an LSP with type checking enabled when writing new code, as it highlights issues with types being assumed where those assumptions may not always hold true (assuming you're using VScode and Pyright, [make sure `python.analysis.typeCheckingMode` is set to something higher than `basic`](https://github.com/microsoft/pyright/blob/main/docs/settings.md))

StandingPadAnimations committed:
refactor: Refactor some of the type handling in main spawner code

Signed-off-by: Maryam Sheikh <mahid@standingpad.org>

StandingPadAnimations committed:
style: Remove random tab character in GPL header

Signed-off-by: Maryam Sheikh <mahid@standingpad.org>

StandingPadAnimations committed:
style: Remove unnecessary dashes in comments

Signed-off-by: Maryam Sheikh <mahid@standingpad.org>

StandingPadAnimations committed:
style: Remove unnecessary whitespace between sections

Signed-off-by: Maryam Sheikh <mahid@standingpad.org>

StandingPadAnimations committed:
style: Consolidate face direction checks

Signed-off-by: Maryam Sheikh <mahid@standingpad.org>

StandingPadAnimations committed:
style: Remove random tab character

Signed-off-by: Maryam Sheikh <mahid@standingpad.org>

StandingPadAnimations reviewed (approved):
Unless there are any objections, I'll merge this into `dev`
